src/jdk.jfr/share/classes/jdk/jfr/internal/tool/Metadata.java

```diff
@@ -91,52 +102,158 @@ int groupValue(Type t) {
+        stream.println();
+    }
+
     @Override
     public void execute(Deque<String> options) throws UserSyntaxException, UserDataException {
-        Path file = getJFRInputFile(options);
+        Path file = getOptionalJFRInputFile(options);
 
         boolean showIds = false;
+        boolean foundEventFilter = false;
+        boolean foundCategoryFilter = false;
+        Predicate<EventType> filter = null;
         int optionCount = options.size();
         while (optionCount > 0) {
-            if (acceptOption(options, "--ids")) {
+            // internal option, doest not export to users
+            if (acceptSingleOption(options, "--ids")) {
                 showIds = true;
             }
+            if (acceptFilterOption(options, "--events")) {
+                if (foundEventFilter) {
+                    throw new UserSyntaxException("use --events event1,event2,event3 to include multiple events");
+                }
+                foundEventFilter = true;
+                String filterStr = options.remove();
+                warnForWildcardExpansion("--events", filterStr);
+                filter = addEventFilter(filterStr, filter);
+            }
+            if (acceptFilterOption(options, "--categories")) {
+                if (foundCategoryFilter) {
+                    throw new UserSyntaxException("use --categories category1,category2 to include multiple categories");
+                }
+                foundCategoryFilter = true;
+                String filterStr = options.remove();
+                warnForWildcardExpansion("--categories", filterStr);
+                filter = addCategoryFilter(filterStr, filter);
+            }
             if (optionCount == options.size()) {
                 // No progress made
+                checkCommonError(options, "--event", "--events");
+                checkCommonError(options, "--category", "--categories");
                 throw new UserSyntaxException("unknown option " + options.peek());
             }
             optionCount = options.size();
         }
 
-        try (PrintWriter pw = new PrintWriter(System.out)) {
+        try (PrintWriter pw = new PrintWriter(System.out, false, Charset.forName("UTF-8"))) {
             PrettyWriter prettyWriter = new PrettyWriter(pw);
             prettyWriter.setShowIds(showIds);
-            try (RecordingFile rf = new RecordingFile(file)) {
-                List<Type> types = PRIVATE_ACCESS.readTypes(rf);
+            if (filter != null) {
+                filter = addCache(filter, eventType -> eventType.getId());
+            }
+
+            // determine whether reading from recording file or reading from the JDK where
+            // the jfr tool is located will be used
+            List<EventType> types = null;
+            if (file != null) {
+                try (RecordingFile rf = new RecordingFile(file)) {
+                    types = rf.readEventTypes();
```

Comment:
Maybe move this part into a separate method and remove the null check.

List<EventType> types = useEvenTypes(file);

private void useEventTypes(Path file) {
    if (file == null) {
       return new ArrayList<?>(FlightRecorder.getEventTypes());
   }
   try (RecordingFile rf = new RecordingFile(file)) {
           return rf.readEventTypes();
   } catch (IOException ioe) {
        couldNotReadError(file, ioe);
   }
}

Reply:
Fixed